src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt

```diff
@@ -0,0 +1,203 @@
+package org.rust.cargo.runconfig.test
+
+import com.intellij.execution.process.ProcessOutputTypes
+import com.intellij.execution.testframework.TestConsoleProperties
+import com.intellij.execution.testframework.sm.ServiceMessageBuilder
+import com.intellij.execution.testframework.sm.runner.OutputToGeneralTestEventsConverter
+import com.intellij.openapi.util.Key
+import jetbrains.buildServer.messages.serviceMessages.ServiceMessage
+import jetbrains.buildServer.messages.serviceMessages.ServiceMessageVisitor
+
+/**
+ * Translate Cargo Test output to TeamCity Service Messages which SM Test Runner understands.
+ *
+ * Unfortunately, as for now Cargo Test only supports `human` message format, which is pretty
+ * obscure to handle programmatically. First of all, it is split into some separate sections,
+ * which are not easily distinguishable, that's why we do state machines here. Next problem is,
+ * that Cargo collects failed tests stdouts and print them yet at the end of test session.
+ * Lastly, we do not have exact information when given test started.
```

Comment:
> I think this can be worked around by test -- --nocapture --test-threads 1.

Nice! This is probably the way to go (as for now).

> Without --test-threads the output can be interleaved in arbitrary ways I think.

Yep, that's probably why rust test is collecting stdout and printing them at the end

> Adding support for custom reporters to Rust tools would be great, but this is more complicated than it seems. The first thing to understand is that you are not interacting with cargo here, but with a test harness binary. I think this is it https://github.com/rust-lang/rust/blob/master/src/libtest/lib.rs, but I haven't worked with it.

I don't see any problems with this. We are already passing some arguments to test binary (note `test -- ...`) and it doesn't look very complicated. So I think extending it shouldn't be a huge problem, especially as not only we would benefit from it.

> There is a consensus that it should be possible to plug a custom test runner instead, which would solve a whole host of problems, including a custom error reporting, but looks like nobody is actively working on it (because this is not a small task!).

Indeed, this is not small task